File: app/src/main/java/site/aiok/onepic/logic/ImageSlicer.kt

package site.aiok.onepic.logic

import android.graphics.Bitmap
import site.aiok.onepic.model.PuzzlePiece
import kotlin.random.Random

object ImageSlicer {

    /**
     * 切分图片为拼图块，并洗牌到随机位置
     * 注意：这里只负责切图和生成块，不设置像素坐标
     * 像素坐标由 GameBoardView.updateLayout() 根据屏幕尺寸计算
     */
    fun sliceImage(originalBitmap: Bitmap, rows: Int, cols: Int): List<PuzzlePiece> {
        val pieces = mutableListOf<PuzzlePiece>()
        val pieceWidth = originalBitmap.width / cols
        val pieceHeight = originalBitmap.height / rows

        var idCounter = 0
        
        for (row in 0 until rows) {
            for (col in 0 until cols) {
                // 原始位置（用于切图）
                val x = col * pieceWidth
                val y = row * pieceHeight

                // 创建此块的 bitmap
                val pieceBitmap = Bitmap.createBitmap(originalBitmap, x, y, pieceWidth, pieceHeight)

                pieces.add(
                    PuzzlePiece(
                        id = idCounter,
                        currentX = 0f, // 将在 updateLayout 中设置
                        currentY = 0f,
                        targetX = 0f,  // 将在 updateLayout 中设置
                        targetY = 0f,
                        width = pieceWidth.toFloat(),
                        height = pieceHeight.toFloat(),
                        bitmap = pieceBitmap,
                        zIndex = 0,
                        groupId = idCounter, // 每块初始独立
                        row = row,           // 目标行（正确位置）
                        col = col            // 目标列（正确位置）
                    )
                )
                idCounter++
            }
        }
        
        return pieces
    }
}

File: ios/OnePic/OnePic/Logic/ImageSlicer.swift
import UIKit
import SwiftUI

struct ImageSlicer {
    
    /// 与 Android Bitmap.createBitmap 一致：按像素坐标切片
    static func slice(image originalImage: UIImage, rows: Int, cols: Int) -> [PuzzlePiece] {
        var pieces: [PuzzlePiece] = []
        
        // 归一化方向后使用 cgImage 像素坐标，与 Android 完全一致
        let image = originalImage.imageOrientation == .up ? originalImage : normalizeOrientation(originalImage)
        guard let cgImage = image.cgImage else {
            print("❌ ImageSlicer: No cgImage after normalize")
            return pieces
        }
        
        let width = cgImage.width
        let height = cgImage.height
        let pieceWidth = width / cols
        let pieceHeight = height / rows
        
        guard pieceWidth > 0, pieceHeight > 0 else {
            print("❌ ImageSlicer: Invalid dimensions \(width)x\(height) for \(rows)x\(cols)")
            return pieces
        }
        
        var idCounter = 0
        
        let bounds = CGRect(x: 0, y: 0, width: width, height: height)
        
        for row in 0..<rows {
            for col in 0..<cols {
                let x = col * pieceWidth
                let y = row * pieceHeight
                var rect = CGRect(x: x, y: y, width: pieceWidth, height: pieceHeight).integral
                rect = rect.intersection(bounds)
                
                // WebP/CGImage.cropping 可能产生黑块，强制用 UIGraphicsImageRenderer 绘制（Android Bitmap 无此问题）
                let pieceImage: UIImage
                if rect.width < 1 || rect.height < 1 {
                    pieceImage = createPlaceholder(width: pieceWidth, height: pieceHeight)
                } else {
                    pieceImage = renderSliceByDraw(from: image, rect: rect, cgImage: cgImage)
                }
                #if DEBUG
                if pieceImage.cgImage == nil {
                    print("⚠️ ImageSlicer: Piece \(idCounter) slice failed cgImage")
                }
                #endif
                
                let piece = PuzzlePiece(
                    id: idCounter,
                    currentX: 0,
                    currentY: 0,
                    targetX: 0,
                    targetY: 0,
                    width: CGFloat(pieceWidth),
                    height: CGFloat(pieceHeight),
                    image: pieceImage,
                    zIndex: 0,
                    isLocked: false,
                    groupId: idCounter,
                    row: row,
                    col: col
                )
                pieces.append(piece)
                idCounter += 1
            }
        }
        
        return pieces
    }
    
    /// 归一化方向，确保 cgImage 与 size 对应
    private static func normalizeOrientation(_ image: UIImage) -> UIImage {
        guard image.imageOrientation != .up else { return image }
        let size = image.size
        let renderer = UIGraphicsImageRenderer(size: size)
        return renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: size))
        }
    }
    
    /// 裁切：优先 CGImage.cropping，失败则用 UIGraphicsImageRenderer（WebP 等格式可能需后者）
    private static func renderSliceByDraw(from image: UIImage, rect: CGRect, cgImage: CGImage) -> UIImage {
        guard rect.width > 0, rect.height > 0 else {
            return createPlaceholder(width: 1, height: 1)
        }
        if let cropped = cgImage.cropping(to: rect) {
            return UIImage(cgImage: cropped, scale: 1, orientation: .up)
        }
        let scale = image.scale
        let format = UIGraphicsImageRendererFormat.default()
        format.scale = 1
        format.opaque = true
        let renderer = UIGraphicsImageRenderer(size: CGSize(width: rect.width, height: rect.height), format: format)
        return renderer.image { _ in
            image.draw(at: CGPoint(x: -rect.origin.x / scale, y: -rect.origin.y / scale))
        }
    }
    
    private static func createPlaceholder(width: Int, height: Int) -> UIImage {
        let w = max(1, width)
        let h = max(1, height)
        let renderer = UIGraphicsImageRenderer(size: CGSize(width: w, height: h))
        return renderer.image { ctx in
            UIColor.darkGray.setFill()
            ctx.fill(CGRect(x: 0, y: 0, width: w, height: h))
        }
    }
}
